elkscmd/elvis/doc/ctags.man
.TH CTAGS 1
.SH NAME
ctags - Generates "tags" and (optionally) "refs" files
.SH SYNOPSIS
\fBctags\fP [\fB-r\fP] \fIfiles\fP...
.SH DESCRIPTION
\fIctags\fP generates the "tags" and "refs" files
from a group of C source files.
The "tags" file is used by Elvis' ":tag" command,
control-] command,
and -t option.
The "refs" file is used by the \fIref(1)\fP program.
.PP
Each C source file is scanned for #define statements and
global function definitions.
The name of the macro or function becomes the name of a tag.
For each tag, a line is added to the "tags" file which contains:
.RS
.nf
       - the name of the tag
       - a tab character
       - the name of the file containing the tag
       - a tab character
       - a way to find the particular line within the file.
.RE
.fi
.PP
The "refs" file is used by the \fIref(1)\fP program, which can be invoked
via Elvis' K command.
When ctags finds a global function definition,
it copies the function header into the "refs" file.
The first line is flush against the right margin,
but the argument definitions are indented.
The ref program can search the "refs" file
.ul
much
faster than it could search all of the C source files.
.PP
The filenames list will typically be the names of all C source
files in the current directory, like this:
.RS
.nf
$ ctags -r *.[ch]
.RE
.fi
.SH OPTIONS
.IP \fB-r\fP
This causes \fIctags\fP to generate both "tags" and "refs".
Without \fB-r\fP, it would only generate "tags".
.SH FILES
.IP tags
The "tags" file.
.IP refs
The "refs" file.
.SH "SEE ALSO"
elvis(1), refs(1)
.SH BUGS
This version of ctags doesn't parse ANSI-C source code very well.
It has trouble recognizing the new-style function definitions.
.SH AUTHOR
.nf
Steve Kirkendall
kirkenda@cs.pdx.edu
\&...uunet!tektronix!psueea!eecs!kirkenda
.fi


elks/CHANGELOG
------------------------------------------------------------------------------
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes.

Al
------------------------------------------------------------------------------
Fri Jul 09 17:00 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Major updates to socket code. Addition of server socket capability,
	  and addition of unix domain socket code (currently no functional),
	  plus nano sockets.

Fri Jul 09 16:59 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Late commit of changes for version 0.0.77 due to cvs outage.

Fri Jul 09 16:54 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Added support for very basic pseudo ttys. Initially to
	  support terminal program for nano-X.

Fri Jul 09 16:21 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Make exit send a SIGCHLD to parent.

Fri Jul 09 16:21 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Fixed the way the kernel handles loading sh if no init is found

Fri Jul 09 16:15 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Updated signal code to make it clearer.

Fri Jul 09 16:10 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Updated romfs to use byteorder functions.

Fri Jul 09 15:45 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Tweaks to bioshd driver to fix problems with disabling hd
	  support. Fix to Changelog times.

Wed Jul 07 19:17 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Updated build system to make it more arch independant. Makefile
	  contributed by Simon Wood.

Tue Jul 06 19:17 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Added byteorder functions for romfs.

Tue Jul 06 14:23 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* More changes to the rd driver by Blaz Antonic.

Mon Jul 05 14:11 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Sanitisation and enhancement of the rd driver by Blaz Antonic.

Thu Jun 24 14:11 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Bug fix to bioscon driver.

Tue Jun 15 10:40 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Latest changes to (ALPHA) meta driver.

Tue Jun 15 10:31 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Minor tweaks to Makefile

Sun May 30 00:05 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Fixes code to allow loading of shared libraries.

Sun May 30 00:05 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Modified tty interface to simplify it.

	* Added support for non-blocking IO on ttys.

	* Included support for graphics on direct console driver.

	* Updated serial termios support to allow baud rate to be changed.

	* Added support for ^C and ^Z to generate signals from the console.

	* Minor tweaks to select code.

	* Added socket to syscall table.

Sun May 30 00:01 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Updated todo file.

Fri May 28 20:04 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Updated bugs file.

Thu May 20 12:04 BST 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Final tweaks for 0.0.76

Mon May 17 12:29 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Added Documentation/cs_write which describes code that
	  writes to the kernel code segment.

Tue May 06 17:19 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Fixed select code, including modifications to scheduler
	  to fix timers.

Tue May 06 16:39 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Minor fix to lp driver.

Fri Apr 16 09:57 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Added support for the signal(2) system call, and support
	  for signal handler functions.

Thu Feb 25 21:13 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Fixed typo for read_only option.

Tue Feb 16 12:34 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Final tweaks for 0.0.75

Mon Feb 15 17:08 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Driver stub of meta driver for user space device drivers

Mon Feb 15 17:06 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Latest tweaks to serial driver by Blaz Antonic

Mon Feb 15 16:56 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>
	
	* Added alpha code for shared libraries.

Mon Feb 15 16:47 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Commented mm diagnostic functions.

Fri Feb 12 16:35 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Trimmed block request code.

Sun Jan 31 17:50 GMT 1999   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Fix main.c so init is launched with a valid set of argvs.

	* Fix ntty.c so tty driver ioctls are called.

Sat Dec 06 00:24 GMT 1998   Alastair Bridgewater  <nyef@softhome.net>

	* Fixed arch/i86/lib/Makefile to rebuild setup[bw].s when nessecary.

Wed Thu 03 17:12 GMT 1998   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Tweaked docs for 0.0.74

Wed Thu 03 17:12 GMT 1998   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Debugged minix filesystem, so that removing files now
	  works correctly, and the filesystem no longer gets corrupted.

Wed Nov 19 18:22 GMT 1998   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Added Riley Williams changes to bootsect.S to fix a bug and tidy
	  up the code.

Wed Nov 19 17:51 GMT 1998   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Added latest changes to 286 pmode code by
	  Alastair Bridgewater. Include change to the code that detects
	  when the kernel is restarted so it no longer has to modify its own
	  code segment.

Wed Nov 19 17:39 GMT 1998   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Fix to minix truncate code. See BUGS.

Tue Nov 18 11:42 GMT 1998   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Tweaked select so it now passes 5 parameters in the normal way.

Thu Nov 18 15:00 GMT 1998   Alistair Riddoch  <ajr@ecs.soton.ac.uk>

	* Some more updates to config scripts made in conjunction with
	  Riley Williams.

Thu Nov 18 14:31 GMT 1998   Alistair Riddoch  <ajr@ecs.soton.ac.uk>


elkscmd/elvis/ctags.c
/* ctags.c */

/* Author:
 *	Steve Kirkendall
 *	14407 SW Teal Blvd. #C
 *	Beaverton, OR 97005
 *	kirkenda@cs.pdx.edu
 */


/* This file contains the complete source to the ctags program. */

/* Special abilities:
 * Can also make a "refs" file for use by the "ref" program.
 */

/* Limitations:
 * This version of ctags always writes its output to the file "tags".
 * It assumes that every command-line argument (but "-r") is a C source file.
 * It does not sort the list of tags, unless CFLAGS=-DSORT.
 * It does not recognize duplicate definitions.
 * It does not try to handle "static" functions in a clever way.
 * It probably won't scan ANSI-C source code very well.
 */

/* Implementation:
 * Lines are scanned one-at-a-time.
 * The context of lines is tracked via a finite state machine.
 * Contexts are:
 *	EXPECTFN - we're looking for a function name.
 *	ARGS	 - between function name and its opening {
 *	BODY	 - we found a function name, skip to end of body.
 *
 * Function tags are referenced by a search string, so that lines may be
 * inserted or deleted without mucking up the tag search.
 *
 * Macro tags are referenced by their line number, because 1) they usually
 * occur near the top of a file, so their line# won't change much; 2)  They
 * often contain characters that are hard to search for; and 3)  Their #define
 * line is likely to be altered.
 *
 * Each line of the resulting "tags" file describes one tag.  Lines begin with
 * the tag name, then a tab, then the file name, then a tab, and then either
 * a line number or a slash-delimited search string.
 */

#include "config.h"

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void ctags(FILE *fp, char *name, FILE *refs);

#define REFS	"refs"

#if OSK
#define NUMFMT	"%%.%ds\t%%s\t%%ld\n"
#define SRCHFMT	"%%.%ds\t%%s\t/^%%s$/\n"
#define MAINFMT	"M%%.%ds\t%%s\t/^%%s$/\n"
static char fmt[256];
#else
#define NUMFMT	"%.*s\t%s\t%ld\n"
#define SRCHFMT	"%.*s\t%s\t/^%s$/\n"
#define MAINFMT	"M%.*s\t%s\t/^%s$/\n"
#endif

#ifdef VERBOSE
# define SAY(x)	fprintf(stderr, "%s\n", x);
#else
# define SAY(x)
#endif

#define EXPECTFN 1
#define	ARGS	 2
#define BODY	 3

extern char	*fgets();

char		*progname;	/* argv[0], used for diagnostic output */

main(argc, argv)
	int	argc;
	char	**argv;
{
	FILE	*fp;
	int	i;
	FILE	*refs;	/* used to write to the refs file */

#if MSDOS || TOS
	char **wildexpand();
	argv=wildexpand(&argc, argv);
#endif
	/* notice the program name */
	progname = argv[0];

	/* create the "refs" file if first arg is "-r" */
	if (argc > 1 && !strcmp(argv[1], "-r"))
	{
		/* delete the "-r" flag from the args list */
		argc--;
		argv++;

		/* open the "refs" file for writing */
		refs = fopen(REFS, "w");
		if (!refs)
		{
			fprintf(stderr, "%s: could not create \"%s\"\n", progname, REFS);
			exit(2);
		}
	}
	else
	{
		refs = (FILE *)0;
	}

	/* process each file named on the command line, or complain if none */
	if (argc > 1)
	{
		/* redirect stdout to go to the "tags" file */
		if (!freopen("tags", "w", stdout))
		{
			fprintf(stderr, "%s: could not create \"%s\"\n", progname, TAGS);
			exit(2);
		}

		for (i = 1; i < argc; i++)
		{
			/* process this named file */
			fp = fopen(argv[i], "r");
			if (!fp)
			{
				fprintf(stderr, "%s: could not read \"%s\"\n", progname, argv[i]);
				continue;
			}
			ctags(fp, argv[i], refs);
			fclose(fp);
		}
#ifdef SORT
		/* This is a hack which will sort the tags list.   It should
		 * on UNIX and Minix.  You may have trouble with csh.   Note
		 * that the tags list only has to be sorted if you intend to
		 * use it with the real vi;  elvis permits unsorted tags.
		 */
		fflush(stdout);
#if OSK
		fclose(stdout);
		system("qsort tags >-_tags; -nx; del tags; rename _tags tags");
#else	
		system("sort tags >_tags$$; mv _tags$$ tags");
#endif
#endif
		exit(0);
	}
	else
	{
		fprintf(stderr, "usage: %s *.[ch]\n", progname);
		exit(2);
	}
}


/*
 * This function finds all tags in a given file
 *
 * fp    stream of the file to scan
 * name  name of the file being scanned
 * refs  NULL, or where to write refs lines
 */

void ctags(FILE *fp, char *name, FILE *refs)
{
	int	context;	/* context - either EXPECTFN, ARGS, or BODY */
	long	lnum;		/* line number */
	char	text[1000];	/* a line of text from the file */
	char	*scan;		/* used for searching through text */
	int	len;		/* length of the line */

	/* for each line of the file... */
	for (context = EXPECTFN, lnum = 1; fgets(text, sizeof text, fp); lnum++)
	{
#ifdef VERBOSE
		switch(context)
		{
		  case EXPECTFN:	scan = "EXPECTFN";	break;
		  case ARGS:		scan = "ARGS    ";	break;
		  case BODY:		scan = "BODY    ";	break;
		  default:		scan = "context?";
		}
		fprintf(stderr, "%s:%s", scan, text);
#endif

		/* start of body? */
		if (text[0] == '{')
		{
			context = BODY;
			SAY("Start of BODY");
			continue;
		}

		/* argument line, to be written to "refs" ? */
		if (refs && context == ARGS)
		{
			if (text[0] != '\t')
			{
				putc('\t', refs);
			}
			fputs(text, refs);
			SAY("Argument line");
			continue;
		}

		/* ignore empty or indented lines */
		if (text[0] <= ' ')
		{
			SAY("Empty or indented");
			continue;
		}

		/* end of body? */
		if (text[0] == '}')
		{
			context = EXPECTFN;
			SAY("End of BODY");
			continue;
		}

		/* ignore lines in the body of a function */
		if (context != EXPECTFN)
		{
			SAY("BODY or ARGS");
			continue;
		}

		/* strip the newline */
		len = strlen(text);
		text[--len] = '\0';

		/* a preprocessor line? */
		if (text[0] == '#')
		{
			/* find the preprocessor directive */
			for (scan = &text[1]; isspace(*scan); scan++)
			{
			}

			/* if it's a #define, make a tag out of it */
			if (!strncmp(scan, "define", 6))
			{
				/* find the start of the symbol name */
				for (scan += 6; isspace(*scan); scan++)
				{
				}

				/* find the length of the symbol name */
				for (len = 1;
				     isalnum(scan[len]) || scan[len] == '_';
				     len++)
				{
				}
#if OSK
				sprintf(fmt, NUMFMT, len);
				printf(fmt, scan, name, lnum);
#else			
				printf(NUMFMT, len, scan, name, lnum);
#endif
			}
			SAY("Preprocessor line");
			continue;
		}

		/* an extern or static declaration? */
		if (text[len - 1] == ';'
		 || !strncmp(text, "extern", 6)
		 || !strncmp(text, "EXTERN", 6)
		 || !strncmp(text, "static", 6)
		 || !strncmp(text, "PRIVATE", 7))
		{
			SAY("Extern or static");
			continue;
		}

		/* if we get here & the first punctuation other than "*" is
		 * a "(" which is immediately preceded by a name, then
		 * assume the name is that of a function.
		 */
		for (scan = text; *scan; scan++)
		{
			if (ispunct(*scan)
			 && !isspace(*scan) /* in BSD, spaces are punctuation?*/
			 && *scan != '*' && *scan != '_' && *scan != '(')
			{
				SAY("Funny punctuation");
				goto ContinueContinue;
			}

			if (*scan == '(')
			{
				/* permit 0 or 1 spaces between name & '(' */
				if (scan > text && scan[-1] == ' ')
				{
					scan--;
				}

				/* find the start & length of the name */
				for (len = 0, scan--;
				     scan >= text && (isalnum(*scan) || *scan == '_');
				     scan--, len++)
				{
				}
				scan++;

				/* did we find a function? */
				if (len > 0)
				{
					/* found a function! */
					if (len == 4 && !strncmp(scan, "main", 4))
					{
#if OSK
						sprintf(fmt, MAINFMT, strlen(name) - 2);
						printf(fmt, name, name, text);
#else			
						printf(MAINFMT, strlen(name) - 2, name, name, text);
#endif
					}
#if OSK
					sprintf(fmt, SRCHFMT, len);
					printf(fmt, scan, name, text);
#else				
					printf(SRCHFMT, len, scan, name, text);
#endif
					context = ARGS;

					/* add a line to refs, if needed */
					if (refs)
					{
						fputs(text, refs);
						putc('\n', refs);
					}

					goto ContinueContinue;
				}
			}
			else
			{
				SAY("No parenthesis");
			}
		}
		SAY("No punctuation");

ContinueContinue:;
	}
}

#if MSDOS || TOS
#define		WILDCARD_NO_MAIN
#include	"wildcard.c"
#endif
